File: src/service/document_service.py
from src.model.document import Document
from src.model.document_type import DocumentType, get_type_by_num
from src.model.type.result import Result
from src.view.bar_view import bar


def print_document(document_list: list[Document]) -> None:
    for i, document in enumerate(document_list):
        print(f"N: {i + 1}\tidentifier: {document.id}")


def get_document_by_type(document_list: list[Document], document_type: DocumentType) -> list[Document]:
    documents = [document for document in document_list if document.type == document_type]
    return documents


def print_document_group_by_type(document_list: list[Document]) -> None:
    cin_list = get_document_by_type(document_list, DocumentType.CIN)
    driving_licence_list = get_document_by_type(document_list, DocumentType.DRIVING_LICENCE)
    bank_card_list = get_document_by_type(document_list, DocumentType.BANK_CARD)

    if cin_list:
        bar("CIN")
        print_document(cin_list)
    if driving_licence_list:
        bar("Driving licence")
        print_document(driving_licence_list)
    if bank_card_list:
        bar("Bank card")
        print_document(bank_card_list)


def add_document(document_list: list[Document]) -> Result:
    document: Document = Document()
    input_type: str = input("Document type:\n\t1. CIN\n\t2. Driving Licence\n\t3. Bank Card\n")
    if input_type.isdigit() and 1 <= int(input_type) <= 3:
        input_identifier = input("Document identifier: ")
        if input_identifier:
            document.type = get_type_by_num(int(input_type))
            document.id = input_identifier
            document_list.append(document)
            return Result(True, "Document added successfully")
        else:
            return Result(False, "Invalid Input - retry")
    else:
        return Result(False, "Invalid Input - retry")


def remove_document_in_list(document_list: list[Document], document: Document) -> None:
    for i in range(len(document_list)):
        if document_list[i] == document:
            document_list.pop(i)
            break


def remove_document(documents: list[Document]) -> Result:
    print_document_group_by_type(documents)
    type_doc_remove_input = input("Type of document you want to remove:\n\t1. CIN\n\t2. Driving Licence\n\t3. Bank "
                                  "Card:\n")
    if type_doc_remove_input.isdigit() and 1 <= int(type_doc_remove_input) <= 3:
        type_doc_remove = get_type_by_num(int(type_doc_remove_input))
        index_doc_remove_input = input("Document number (N): ")
        if index_doc_remove_input.isdigit():
            index_doc_remove = int(index_doc_remove_input)
            documents_by_type = get_document_by_type(documents, type_doc_remove)
            if len(documents_by_type) >= index_doc_remove:
                document_removed = documents_by_type[index_doc_remove - 1]
                remove_document_in_list(documents, document_removed)
                return Result(True, "Document removed successfully")
            else:
                return Result(False, "Invalid Input - retry")
    else:
        return Result(False, "Invalid Input - retry")

File: lib/service/document_service.rb
require_relative '../model/document'
require_relative '../model/document_type'
require_relative '../model/types/result'
require_relative '../view/bar_view'

def print_document(document_list)
  document_list.each_with_index do |document, i|
    puts "N: #{i + 1}\tidentifier: #{document.id}"
  end
end

def get_document_by_type(document_list, document_type)
  document_list.select { |document| document.type == document_type }
end

def print_document_group_by_type(document_list)
  cin_list = get_document_by_type(document_list, DocumentType::CIN)
  driving_licence_list = get_document_by_type(document_list, DocumentType::DRIVING_LICENCE)
  bank_card_list = get_document_by_type(document_list, DocumentType::BANK_CARD)

  bar("CIN") if cin_list.any?
  print_document(cin_list) if cin_list.any?

  bar("Driving Licence") if driving_licence_list.any?
  print_document(driving_licence_list) if driving_licence_list.any?

  bar("Bank Card") if bank_card_list.any?
  print_document(bank_card_list) if bank_card_list.any?
end

def add_document(document_list)
  document = Document.new
  puts "Document type:\n\t1. CIN\n\t2. Driving Licence\n\t3. Bank Card\n"
  input_type = gets.chomp
  if input_type.match?(/\A\d+\z/) && (1..3).cover?(input_type.to_i)
    puts "Document identifier: "
    input_identifier = gets.chomp
    if !input_identifier.empty?
      document.id = input_identifier
      document.type = type_from_number(input_type.to_i)
      document_list << document
      Result.new(true, "Document added successfully")
    else
      Result.new(false, "Invalid Input - retry")
    end
  else
    Result.new(false, "Invalid Input - retry")
  end
end

def remove_document_in_list(document_list, document)
  document_list.delete(document)
end

def remove_document(documents)
  print_document_group_by_type(documents)
  puts "Type of document you want to remove:\n\t1. CIN\n\t2. Driving Licence\n\t3. Bank Card:\n"
  type_doc_remove_input = gets.chomp
  if type_doc_remove_input.match?(/\A\d+\z/) && (1..3).cover?(type_doc_remove_input.to_i)
    type_doc_remove = type_from_number(type_doc_remove_input.to_i)
    puts "Document number (N): "
    index_doc_remove_input = gets.chomp
    if index_doc_remove_input.match?(/\A\d+\z/)
      index_doc_remove = index_doc_remove_input.to_i
      documents_by_type = get_document_by_type(documents, type_doc_remove)
      if documents_by_type.length >= index_doc_remove
        document_removed = documents_by_type[index_doc_remove - 1]
        remove_document_in_list(documents, document_removed)
        Result.new(true, "Document removed successfully")
      else
        Result.new(false, "Invalid Input - retry")
      end
    else
      Result.new(false, "Invalid Input - retry")
    end
  else
    Result.new(false, "Invalid Input - retry")
  end
end
